File: rust/src/clock.rs
//! Wall and elapsed clocks.

use std::time::Duration;

use futures::future::BoxFuture;

/// Separates wall timestamps from monotonic elapsed time and timers.
///
/// Wall time stamps frames and invalidation markers. Elapsed time governs
/// deadlines and process-local expiry. Timers deliver deadline callbacks; a
/// controlled clock may deliver them on its own schedule.
pub trait Clock: Send + Sync + 'static {
    /// Epoch milliseconds from the application wall clock.
    fn wall_ms(&self) -> i64;
    /// Monotonic elapsed time since this clock's origin, with native precision.
    fn elapsed(&self) -> Duration;
    /// Whole-millisecond elapsed reading used by process-local storage.
    ///
    /// Insertion and lookup share this integer grid, so an entry inserted at
    /// 0.7 ms with a 1,000 ms TTL expires when this reading reaches 1,000.
    fn elapsed_ms(&self) -> i64 {
        self.elapsed().as_millis().min(i64::MAX as u128) as i64
    }
    /// A timer that completes once `duration` of this clock's time has passed.
    fn sleep(&self, duration: Duration) -> BoxFuture<'static, ()>;
}

#[cfg(feature = "tokio")]
mod system {
    use super::Clock;
    use futures::future::BoxFuture;
    use std::sync::OnceLock;
    use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

    static PROCESS_ORIGIN: OnceLock<Instant> = OnceLock::new();

    /// The default clock: system wall time, a monotonic origin aligned to the
    /// process-wide millisecond grid, and tokio timers.
    #[derive(Debug, Clone)]
    pub struct SystemClock {
        origin: Instant,
    }

    impl SystemClock {
        /// A clock whose whole-millisecond grid is shared by every default instance.
        pub fn new() -> Self {
            let process = *PROCESS_ORIGIN.get_or_init(Instant::now);
            let now = Instant::now();
            let phase = now.saturating_duration_since(process).as_nanos() % 1_000_000;
            let origin = now.checked_sub(Duration::from_nanos(phase as u64)).unwrap_or(now);
            SystemClock { origin }
        }
    }

    impl Default for SystemClock {
        fn default() -> Self {
            Self::new()
        }
    }

    impl Clock for SystemClock {
        fn wall_ms(&self) -> i64 {
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .map(|d| d.as_millis().min(i64::MAX as u128) as i64)
                .unwrap_or(0)
        }

        fn elapsed(&self) -> Duration {
            self.origin.elapsed()
        }

        fn sleep(&self, duration: Duration) -> BoxFuture<'static, ()> {
            Box::pin(tokio::time::sleep(duration))
        }
    }
}

#[cfg(feature = "tokio")]
pub use system::SystemClock;

File: python/dialcache/clock.py
"""Separate wall timestamps, monotonic elapsed time, and timer delivery."""

from __future__ import annotations

import asyncio
import time
from collections.abc import Callable
from typing import Protocol


class TimerHandle(Protocol):
    def cancel(self) -> None: ...


class Clock(Protocol):
    """Injectable clocks and timers for deterministic application tests.

    Wall time stamps Redis frames. Monotonic time governs local expiry and
    deadlines. A controlled clock may advance elapsed time without delivering
    timers, so callers also check elapsed time when operations settle.
    """

    def wall_ms(self) -> int | float: ...

    def monotonic_ms(self) -> int | float: ...

    def call_later(self, milliseconds: float, callback: Callable[[], None]) -> TimerHandle: ...


class SystemClock:
    """System wall time and the process-wide monotonic millisecond grid."""

    def wall_ms(self) -> int:
        return time.time_ns() // 1_000_000

    def monotonic_ms(self) -> float:
        return time.monotonic_ns() / 1_000_000

    def call_later(self, milliseconds: float, callback: Callable[[], None]) -> TimerHandle:
        return asyncio.get_running_loop().call_later(milliseconds / 1_000, callback)

    async def sleep_ms(self, milliseconds: float) -> None:
        future: asyncio.Future[None] = asyncio.get_running_loop().create_future()

        def wake() -> None:
            if not future.done():
                future.set_result(None)

        timer = self.call_later(milliseconds, wake)
        try:
            await future
        finally:
            timer.cancel()
